### PR TITLE
Non-invalidating views

### PR DESCRIPTION
**Pull request by rsinger**
This change lets view spec TTLs be `-1` which essentially means they never change unless the original graph is deleted or changes its rdf type.

### CODE REVIEW

File: src/mongo/delegates/Views.class.php [hunk 1]
```diff
@@ -479,15 +500,19 @@ public function generateView($viewId,$resource=null,$context=null,$queueName=nul
                 $buildImpactIndex=true;
                 if (isset($viewSpec['ttl'])) {
                     $buildImpactIndex=false;
-                    $value[_EXPIRES] = \Tripod\Mongo\DateUtil::getMongoDate($this->getExpirySecFromNow($viewSpec['ttl']) * 1000);
+                    if (is_int($viewSpec['ttl']) && $viewSpec['ttl'] > 0) {
+                        $value[_EXPIRES] = \Tripod\Mongo\DateUtil::getMongoDate(
+                            $this->getExpirySecFromNow($viewSpec['ttl']) * 1000
+                        );
+                    }
```
Comment (camillef) [inline src/mongo/delegates/Views.class.php]: So this bit is the main change?

File: test/unit/mongo/MongoTripodViewsTest.php [hunk 2]
```diff
@@ -521,6 +522,116 @@ public function testGenerateViewWithTTL()
...
+                        'dct:references' => ['u' => 'http://talisaspire.com/resources/1234'],
+                        'dct:created' => ['l' => '2018-04-09T00:00:00Z'],
+                        'dct:title' => ['l' => 'A significant event']
+                    ],
+                    [
+                        '_id' => [
+                            'r' => 'http://talisaspire.com/resources/1234',
+                            'c' => 'http://talisaspire.com/'
+                        ],
+                        'dct:title' => ['l' => 'A real piece of work'],
```
Comment (camillef) [inline test/unit/mongo/MongoTripodViewsTest.php]: Yeesh

**Comment by rsinger** [comment]
@camillef It doesn't change publish list performance, but what it *does* do prevent all of the changes done to draft lists from affecting published list related views, because it removes the impact index.

File: src/mongo/delegates/Views.class.php (hunk 1 above)
Comment (rsinger) [reply src/mongo/delegates/Views.class.php]: And lines 82 - 113.

File: src/mongo/delegates/Views.class.php [hunk 3]
```diff
@@ -11,6 +11,7 @@
 use \MongoDB\Driver\ReadPreference;
 use \MongoDB\Collection;
 use Tripod\Mongo\JobGroup;
+use Monolog\Logger;
```
Comment (rgubby) [inline src/mongo/delegates/Views.class.php]: This isn't used

File: src/mongo/delegates/Views.class.php [hunk 4]
```diff
@@ -292,11 +311,13 @@ public function generateViews($resources,$context=null)
         foreach ($resources as $resource)
         {
             $resourceAlias = $this->labeller->uri_to_alias($resource);
-
+            $this->getLogger()->warning('Generating views', ['store' => $this->storeName,
+            '_id' => $resourceAlias]);
```
Comment (rgubby) [inline src/mongo/delegates/Views.class.php]: this formatting looks out of alignment

File: test/unit/mongo/MongoTripodConfigTest.php [hunk 5]
```diff
@@ -973,23 +973,28 @@ public function testSearchConfigNotPresent()
...
-            "resourcelist:List",
-            "spec:User",
-            "bibo:Document",
-            "baseData:Wibble",
-            "baseData:DocWithSequence"
+        $types = $this->tripodConfig->getAllTypesInSpecifications('tripod_php_testing');
+        $this->assertEquals(
+            12,
+            count($types),
+            'There should be 11 types based on the configured view, table and search specifications in config.json'
```
Comment (rgubby) [inline test/unit/mongo/MongoTripodConfigTest.php]: there should be 12.

File: test/unit/mongo/MongoTripodViewsTest.php [hunk 6]
```diff
@@ -6,6 +6,7 @@
 
 use \Tripod\Mongo\Composites\Views;
```
Comment (rgubby) [inline test/unit/mongo/MongoTripodViewsTest.php]: This has never been used and can be removed.

File: test/unit/mongo/MongoTripodViewsTest.php [hunk 7]
```diff
@@ -521,6 +522,116 @@ public function testGenerateViewWithTTL()
...
 
+    public function testNonExpiringViewWithNegativeTTL()
+    {
+        $views = new \Tripod\Mongo\Composites\Views(
+            $this->viewsConstParams[0],
+            $this->viewsConstParams[1],
+            $this->viewsConstParams[2]
+        );
+
+        $view = $views->getViewForResource(
```
Comment (rgubby) [inline test/unit/mongo/MongoTripodViewsTest.php]: This isn't used - should it be tested?

File: test/unit/mongo/MongoTripodViewsTest.php [hunk 8]
```diff
@@ -521,6 +522,116 @@ public function testGenerateViewWithTTL()
...
+            'http://talisaspire.com/resources/1234',
+            'http://purl.org/dc/terms/title',
+            'A real piece of work',
+            'A literal treasure'
+        );
+
+        // This should not affect the view at all
+        $this->tripod->saveChanges($graph, $updatedGraph);
+
+        $view = $views->getViewForResource(
```
Comment (rgubby) [inline test/unit/mongo/MongoTripodViewsTest.php]: This isn't used - should it be tested?

File: test/unit/mongo/MongoTripodViewsTest.php (hunk 7 above)
Comment (rsinger) [reply test/unit/mongo/MongoTripodViewsTest.php]: It's used indirectly, that's how we ensure that the view has actually been generated.

**Pull request merged**